### PR TITLE
[mutable-arrays] add fancy transpose rules for slice / dynamic_slice

### PR DESCRIPTION
**Pull request by mattjj**
This enables users to get sparse updating of gradient accumulators even when differentiating functions on values, e.g.:

```python
@jax.jit
def f(x, i):
  return x[i] ** 2

x = jnp.arange(10.)
grad_accum = jax.new_ref(jnp.zeros(10))

_, f_vjp = jax.vjp(lambda x: f(x, 5), x)
f_vjp = f_vjp.with_refs(grad_accum)
f_vjp(1.)
```

without the `with_refs` line, the backward pass would look like

```
lambda ; b:i32[] c:f32[] k:f32[]. let
  l:f32[] = mul k c
  m:f32[1] = broadcast_in_dim[
    broadcast_dimensions=()
    shape=(1,)
    sharding=None
  ] l
  n:f32[10] = broadcast_in_dim[
    broadcast_dimensions=()
    shape=(10,)
    sharding=None
  ] 0.0:f32[]
  o:f32[10] = dynamic_update_slice n m b
in (o,) }
```

but now, with `with_refs` and after this PR, we get a backward pass that looks like

```
lambda ; c:i32[] d:f32[] b:Ref{f32[10]} l:f32[]. let
  m:f32[] = mul l d
  n:f32[1] = broadcast_in_dim[
    broadcast_dimensions=()
    shape=(1,)
    sharding=None
  ] m
  b[c:c+1] += n
in () }
```

### CODE REVIEW

**Review by yashk2810 — APPROVED** [review]

File: jax/_src/lax/slicing.py [hunk 1]
```diff
@@ -1431,6 +1432,25 @@ def _slice_transpose_rule(t, operand, *, start_indices, limit_indices, strides):
...
+    if strides is None or np.all(np.equal(strides, 1)):
+      pads = zip(start_indices, np.subtract(operand.aval.shape, limit_indices),
+                 (0,) * len(start_indices))
+    else:
+      real_limits = np.add(
+        start_indices,
+        np.where(np.array(out_ct.shape) == 0, 0,
+                 np.add(1, np.multiply(np.subtract(out_ct.shape, 1), strides))))
+      pads = zip(start_indices, np.subtract(operand_shape, real_limits),
+                 np.subtract(strides, 1))
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/_src/lax/slicing.py]: Thank you!

File: jax/_src/lax/slicing.py [hunk 2]
```diff
@@ -1542,6 +1563,18 @@ def _dynamic_slice_transpose_rule(t, operand, *start_indices, slice_sizes):
     return ([dynamic_update_slice_p.bind(zeros, t, *start_indices)] +
             [None] * len(start_indices))
 
+def _dynamic_slice_transpose_fancy(out_ct, operand, *start_indices, slice_sizes):
+  assert isinstance(operand, ad.GradAccum)
+  assert all(not isinstance(s, ad.GradAccum) for s in start_indices)
+  if type(out_ct) is ad_util.Zero: return
```
[the remark this answers is not in the document]
Comment (mattjj) [reply jax/_src/lax/slicing.py:1569]: Subclasses are bad!
